synapse/storage/controllers/state.py

```diff
@@ -501,6 +502,31 @@ async def get_canonical_alias_for_room(self, room_id: str) -> Optional[str]:
 
         return event.content.get("alias")
 
+    @cached()
+    async def get_server_acl_for_room(
+        self, room_id: str
+    ) -> Optional[ServerAclEvaluator]:
+        """Get the server ACL evaluator for room, if any
+
+        This does up-front parsing of the content to ignore bad data and pre-compile
+        regular expressions.
+
+        Args:
+            room_id: The room ID
+
+        Returns:
+            The server ACL evaluator, if any
+        """
+
+        acl_event = await self.get_current_state_event(
+            room_id, EventTypes.ServerACL, ""
+        )
```

Comment:
If you wanted to avoid the call from data store -> storage controller, then there are two options:

1. Push this code down into the datastore (you'd have to use `get_partial_current_state_ids` family of functions), and call `await_full_state` before then. 
2. Add a call to the storage controllers to handle cache invalidation, like we do for data stores, here: https://github.com/matrix-org/synapse/blob/2763c49eca483dbb848b70b951891afd57016f17/synapse/replication/tcp/client.py#L107-L110

Reply:
Ah right, this is slightly annoying because there's 3 spots that need to have similar code here:

* `on_rdata`
* Persisting local events
* Persisting federated events